bftengine/src/bftengine/ReplicaImp.cpp

```diff
@@ -69,58 +70,86 @@ using namespace concord::diagnostics;
 namespace bftEngine::impl {
 
 void ReplicaImp::registerMsgHandlers() {
-  msgHandlers_->registerMsgHandler(MsgCode::Checkpoint, bind(&ReplicaImp::messageHandler<CheckpointMsg>, this, _1));
+  msgHandlers_->registerMsgHandler(MsgCode::Checkpoint,
+                                   bind(&ReplicaImp::messageHandler<CheckpointMsg>, this, _1),
+                                   bind(&ReplicaImp::validatedMessageHandler<CheckpointMsg>, this, _1));
```

Comment:
nit:
I assume you are registering here some messages which will never be handled by the handler?
Another option would be to have a null default value as 2nd parameter for registerMsgHandler which signals "I do not wish to register this message, it is too short to validate async..".

example:

```
msgHandlers_->registerMsgHandler(MsgCode::Checkpoint,
                                   bind(&ReplicaImp::messageHandler<CheckpointMsg>, this, _1),
                                   CheckpointMsg::shouldValidateAsync() ? 
                                   bind(&ReplicaImp::validatedMessageHandler<CheckpointMsg>, this, _1) : nullptr);
);
```

comments:

- It might look uglier, but maybe with more thinking there is a better way.
- You will need to make shouldValidateAsync static, whic hmakes sense.

Reply:
Thanks @glevkovich, this is a great comment to discuss and then finalize.
Comment 1:
I thought about it, to add the handler for specific messages. But later I thought that it might increase an extra step when we would like to add a new message for asynchronous validation. If we add all message then choice of adding or not can happen through shouldValidateAsync(). So I preferred this way.
Comment 2:
Yes we can make it async, but sometimes for some complex message, it might be possible that whether to do async validation or not, will depend on some other cases. e,g, there are many kind of client req messages. Some might not be originating from clients and this may not have any signature. So we can decide case by case whether for this message any asynchronous validation is needed or not.